src/main/java/org/cbioportal/service/FrontendPropertiesServiceImpl.java

```diff
@@ -291,18 +295,63 @@ private Map<String,String> cloneProperties() {
             );
     }
 
-    private String readFile(String fileName) {
-        if (fileName != null && !fileName.isEmpty()) {
-            try (BufferedReader br = Files.newBufferedReader(Paths.get(fileName))) {
-                return br.lines().map(String::trim).collect(Collectors.joining(""));
-            } catch (Exception e) {
-                log.error("Error reading frontend config file: {}", e.getMessage());
-                return null;
+    /**
+     * Read the file and return the content as a single-line string. This works for:
+     * 1) loose files given absolute path
+     * 2) loose files relative to PORTAL_HOME
+     * 3) "classpath:..."" in the app.jar (relative to PORTAL_HOME)
+     * @propertiesFileName: the file path 
+     * REF: based on getResourceStream() in WebServletContextListener.java
+     */
+    private String readFile(String propertiesFileName) {
+        if (propertiesFileName == null || propertiesFileName.isEmpty()) {
+            return null;
+        }
+
+        InputStream inputStream = null;
+
+        // strip off classpath prefix and always check ClassLoader and file system, with latter taking precedence
+        String filePath = propertiesFileName.startsWith("classpath:") 
+            ? propertiesFileName.substring("classpath:".length()) 
+            : propertiesFileName;
+
+        try {
+            String home = System.getenv("PORTAL_HOME");
+
+            // try absolute or relative to working directory
+            File file = new File(filePath);
+            if (file.exists()) {
+                inputStream = new FileInputStream(file);
+            } else if (home != null) {
+                // try relative to PORTAL_HOME
+                file = new File(Paths.get(home, filePath).toString());
+                if (file.exists()) {
+                    inputStream = new FileInputStream(file);
+                }
+            } 
+
+            if (inputStream != null) {
+                log.info("Reading frontend config file: {}", file.getAbsolutePath());
+            } else {
+                // try resource (e.g. app.jar)
+                inputStream = this.getClass().getClassLoader().getResourceAsStream(filePath);
+                if (inputStream != null) {
+                    log.info("Reading frontend config resource: {}", filePath);
+                } else {
+                    throw new Exception("File not found in system or classpath: " + filePath);
+                }
             }
+
+            // read the file and return the content as a single-line string
+            // PRE: inputStream != null
+            BufferedReader br = new BufferedReader(new InputStreamReader(inputStream));
+            return br.lines().map(String::trim).collect(Collectors.joining(""));            
```

Comment:
```suggestion
         Optional<File> file = attemptToFindFile(filePath);
         try (BufferedReader br = Files.newBufferedReader(Paths.get(fileName))) {
                return br.lines().map(String::trim).collect(Collectors.joining(""));
            } catch (Exception e) {
                log.error("Error reading frontend config file: {}", e.getMessage());
                return null;
            }
```

Good catch. Can we wrap the logic into a separate function that is responsible for checking if the file exists?

Reply:
Suggestion doesn't build, just an example

Reply:
Yes, that's a good idea since the function is long. Will update.